File: backend/app/services/diagnostic_cycles.py
"""Identify protocol diagnostic/support cycles for the portable report.

This mirrors the frontend cycle-plot filter. It uses the lower of the charge
and discharge capacities for each cycle, compares that value with a local
post-formation median for the same series, and hides lower-tail outliers. The
rule is intentionally generic: it does not depend on DCIR protocol
recognition, and the report keeps the complete underlying result.
"""
from __future__ import annotations

import math
from statistics import median

DIAGNOSTIC_SIGNALS = ("charge_capacity_mah", "discharge_capacity_mah")
DEFAULT_WINDOW = 21
DEFAULT_TOLERANCE = 0.25
DEFAULT_MIN_CYCLES = 12
DEFAULT_FORMATION_CYCLES = 0


def _finite_positive(value: object) -> float | None:
    if not isinstance(value, (int, float)):
        return None
    number = float(value)
    return number if math.isfinite(number) and number > 0 else None


def find_diagnostic_cycles(
    cycles: list[int],
    capacities: list[float | None],
    *,
    window: int = DEFAULT_WINDOW,
    tolerance: float = DEFAULT_TOLERANCE,
    min_cycles: int = DEFAULT_MIN_CYCLES,
    formation_cycles: int = DEFAULT_FORMATION_CYCLES,
) -> set[int]:
    """Return lower-capacity outliers against a local post-formation median."""
    flagged: set[int] = set()
    post_formation_count = sum(cycle > formation_cycles for cycle in cycles)
    if post_formation_count < min_cycles:
        return flagged

    half = max(1, window // 2)
    for index, cycle in enumerate(cycles):
        if cycle <= formation_cycles:
            continue
        value = _finite_positive(capacities[index] if index < len(capacities) else None)
        if value is None:
            continue
        neighbourhood = []
        for neighbour_index in range(
            max(0, index - half), min(len(capacities), index + half + 1)
        ):
            if cycles[neighbour_index] <= formation_cycles:
                continue
            other = _finite_positive(capacities[neighbour_index])
            if other is not None:
                neighbourhood.append(other)
        if not neighbourhood:
            continue
        local = median(neighbourhood)
        if local > 0 and value < local * (1 - tolerance):
            flagged.add(cycle)
    return flagged


def find_in_series(
    series: dict,
    *,
    tolerance: float = DEFAULT_TOLERANCE,
    formation_cycles: int = DEFAULT_FORMATION_CYCLES,
) -> set[int]:
    quantities = series.get("quantities") or {}
    charge = quantities.get(DIAGNOSTIC_SIGNALS[0])
    discharge = quantities.get(DIAGNOSTIC_SIGNALS[1])
    cycles = series.get("x") or []
    if not charge or not discharge:
        return set()

    capacities: list[float | None] = []
    for index in range(len(cycles)):
        charge_capacity = _finite_positive(charge[index] if index < len(charge) else None)
        discharge_capacity = _finite_positive(
            discharge[index] if index < len(discharge) else None
        )
        capacities.append(
            min(charge_capacity, discharge_capacity)
            if charge_capacity is not None and discharge_capacity is not None
            else None
        )
    return find_diagnostic_cycles(
        cycles,
        capacities,
        tolerance=tolerance,
        formation_cycles=formation_cycles,
    )


def find_across(
    result: dict,
    *,
    tolerance: float = DEFAULT_TOLERANCE,
    formation_cycles: int = DEFAULT_FORMATION_CYCLES,
) -> list[int]:
    """Union of diagnostic cycles across the plotted series, sorted."""
    flagged: set[int] = set()
    for series in result.get("cell_series") or []:
        if series.get("excluded"):
            continue
        flagged |= find_in_series(
            series,
            tolerance=tolerance,
            formation_cycles=formation_cycles,
        )
    return sorted(flagged)


def cycle_ranges(cycles: list[int]) -> list[tuple[int, int]]:
    ranges: list[tuple[int, int]] = []
    for cycle in sorted(set(cycles)):
        if ranges and cycle == ranges[-1][1] + 1:
            ranges[-1] = (ranges[-1][0], cycle)
        else:
            ranges.append((cycle, cycle))
    return ranges


def format_ranges(cycles: list[int]) -> str:
    """Compact, auditable summary of hidden cycle runs."""
    return ", ".join(
        str(start) if start == end else f"{start}–{end}"
        for start, end in cycle_ranges(cycles)
    )

File: frontend/src/features/analyses/editor/families/cycles/diagnosticCycles.ts
/**
 * Identify protocol diagnostic/support cycles after the fact.
 *
 * Diagnostic blocks contain supporting steps around the actual pulse. Those
 * cycles are not ordinary battery cycling and should be hidden together. The
 * detector therefore uses the lower of the charge and discharge capacities,
 * rather than trying to recognize a particular DCIR protocol.
 *
 * Each cell supplies its own baseline. After formation, a cycle is hidden when
 * its lower phase capacity is more than the configured tolerance below the
 * median of its neighbouring post-formation cycles. This remains a
 * presentation-level filter: stored and exported scientific data are kept.
 */

export interface DiagnosticCycleOptions {
  /** Cycles in the centred window used for the local reference. */
  window?: number;
  /**
   * Lower-tail deviation from the local median that marks a cycle as
   * diagnostic. 0.25 means "below 75% of the local median".
   */
  tolerance?: number;
  /** Leave short series alone; there is no reliable local baseline. */
  minCycles?: number;
  /** Number of initial formation cycles excluded from the baseline and filter. */
  formationCycles?: number;
}

export const DIAGNOSTIC_DEFAULTS: Required<DiagnosticCycleOptions> = {
  window: 21,
  tolerance: 0.25,
  minCycles: 12,
  formationCycles: 0,
};

/** The two phase capacities are combined by taking the lower value. */
export const DIAGNOSTIC_SIGNALS = [
  "charge_capacity_mah",
  "discharge_capacity_mah",
] as const;

function isFinitePositive(value: number | null | undefined): value is number {
  return value !== null && value !== undefined && Number.isFinite(value) && value > 0;
}

function median(values: number[]): number | null {
  if (values.length === 0) return null;
  const sorted = [...values].sort((a, b) => a - b);
  const middle = sorted.length >> 1;
  return sorted.length % 2 === 0 ? (sorted[middle - 1] + sorted[middle]) / 2 : sorted[middle];
}

/**
 * Return cycle numbers whose lower phase capacity is below the local baseline.
 *
 * `capacities` must be aligned with `cycles`. Nulls and non-positive values are
 * treated as unknown rather than evidence; a cycle is only judged when both
 * phase capacities were available to form the lower-capacity signal.
 */
export function findDiagnosticCycles(
  cycles: number[],
  capacities: (number | null)[],
  options: DiagnosticCycleOptions = {},
): Set<number> {
  const { window, tolerance, minCycles, formationCycles } = {
    ...DIAGNOSTIC_DEFAULTS,
    ...options,
  };
  const flagged = new Set<number>();
  const postFormationCount = cycles.filter((cycle) => cycle > formationCycles).length;
  if (postFormationCount < minCycles) return flagged;

  const half = Math.max(1, Math.floor(window / 2));
  for (let i = 0; i < cycles.length; i += 1) {
    if (cycles[i] <= formationCycles) continue;
    const value = capacities[i];
    if (!isFinitePositive(value)) continue;

    const neighbourhood: number[] = [];
    for (let j = Math.max(0, i - half); j <= Math.min(capacities.length - 1, i + half); j += 1) {
      if (cycles[j] <= formationCycles) continue;
      const other = capacities[j];
      if (isFinitePositive(other)) neighbourhood.push(other);
    }
    const local = median(neighbourhood);
    if (local === null || local <= 0) continue;
    if (value < local * (1 - tolerance)) flagged.add(cycles[i]);
  }
  return flagged;
}

/** Diagnostic cycles in one series, judged on the lower phase capacity. */
export function findDiagnosticCyclesInSeries(
  series: { x: number[]; quantities: Record<string, (number | null)[]> },
  options: DiagnosticCycleOptions = {},
): Set<number> {
  const charge = series.quantities?.[DIAGNOSTIC_SIGNALS[0]];
  const discharge = series.quantities?.[DIAGNOSTIC_SIGNALS[1]];
  if (!charge || !discharge) return new Set();

  const capacities = series.x.map((_, index) => {
    const chargeCapacity = charge[index];
    const dischargeCapacity = discharge[index];
    if (!isFinitePositive(chargeCapacity) || !isFinitePositive(dischargeCapacity)) return null;
    return Math.min(chargeCapacity, dischargeCapacity);
  });
  return findDiagnosticCycles(series.x, capacities, options);
}

/** Union of diagnostic cycles across every series being plotted. */
export function findDiagnosticCyclesAcross(
  series: { x: number[]; quantities: Record<string, (number | null)[]> }[],
  options: DiagnosticCycleOptions = {},
): Set<number> {
  const all = new Set<number>();
  for (const one of series) {
    for (const cycle of findDiagnosticCyclesInSeries(one, options)) all.add(cycle);
  }
  return all;
}

/** Collapse a sorted cycle list into contiguous runs. */
export function cycleRanges(cycles: number[]): [number, number][] {
  const sorted = [...new Set(cycles)].sort((a, b) => a - b);
  const ranges: [number, number][] = [];
  for (const cycle of sorted) {
    const last = ranges[ranges.length - 1];
    if (last && cycle === last[1] + 1) last[1] = cycle;
    else ranges.push([cycle, cycle]);
  }
  return ranges;
}

/** Render hidden cycles compactly. */
export function formatCycleRanges(cycles: number[], limit = Infinity): string {
  const ranges = cycleRanges(cycles);
  const shown = ranges.slice(0, limit);
  const text = shown.map(([a, b]) => (a === b ? `${a}` : `${a}–${b}`)).join(", ");
  const rest = ranges.length - shown.length;
  return rest > 0 ? `${text}, and ${rest} more` : text;
}

export interface DiagnosticSummary {
  hidden: number[];
  hiddenCount: number;
  shownCount: number;
  ranges: [number, number][];
}

/** Report both sides of the split: what was removed and what remains. */
export function summarizeHidden(allCycles: number[], hidden: Set<number>): DiagnosticSummary {
  const present = [...new Set(allCycles)];
  const removed = present.filter((cycle) => hidden.has(cycle)).sort((a, b) => a - b);
  return {
    hidden: removed,
    hiddenCount: removed.length,
    shownCount: present.length - removed.length,
    ranges: cycleRanges(removed),
  };
}
